### ISSUE TITLE
502 Bad Gateway when trying to proxy to Azure

### ISSUE THREAD

**Issue by kolomik** [damianh/ProxyKit#179]
Hello,

Your solution is the best and can proxy everything, including web sockets. But it doesn't work when proxying to Azure Web Services: when trying to load any resource with GET request server responds with 502 Bad Gateway. As our project is is already running in Azure (and I assume, many others),  we've made an attempt to fix an issue. It's simple adjustment and already was made into PR to your repo: https://github.com/damianh/ProxyKit/pull/178

This fix proved well with real request proxying, using .Net Core 3.0 Web App. But after running tests we've found that in test environments **Request.ContentLength** is always **_null_** and therefore tests are failing. 

Could you consider adjusting either your tests or maybe propose better fix for this situation? Reason of failure for Azure is because Content is set and sent for every request, even if original request had no content (in that case content is still present, but it's empty, I presume). And Azure is not accepting GET requests with body.

For now we have to use forked version of your repo with failing "body tests".

**Comment by pksorensen** [comment]
I can confirm the same thing, my application works on proxykit 2.1.3 and when updating to 2.2.1, i see the same result.

I have not been able to figure out form commits, why that it broke yet.

**Comment by kolomik** [comment]
@pksorensen: Changes for this issue https://github.com/damianh/ProxyKit/issues/168 are causing problems. And it was released in 2.2.1, yes

**Comment by abergs** [comment]
Same here!

**Comment by osoykan** [comment]
Hi, 

There are examples that also need to keep the body as is. 

So, maybe introducing a conditioning mechanism to `ProxyOptions` doing the selection based copying on that could have been a solution. Otherwise putting any pre-decided `ifs` to this section would bother someone else since everyone has different needs apparently. 

I created a PR for that maybe we can discuss if it fits your solution. 
https://github.com/damianh/ProxyKit/pull/184

**Comment by kolomik** [comment]
@osoykan this looks like perfect solution to please everyone and grant more control in usage.

Thank you. Looking forward for new version

**Comment by damianh** [comment]
My philosophy would be for things to just work if at all possible without burdening the user with additional configuration, and have to figure out in what situations to flip it on/off.

> Azure is not accepting GET requests with body

That's annoying, it should just be ignoring it imho.

> because Content is set and sent for every request, even if original request had no content

I think this is the area we need to look at. The original pr that checked the `Content-Length` appeals to me - seems like reasonable course of action.

> There are examples that also need to keep the body as is.

@osoykan If the incoming request has no content then things should work if the is no (empty) body on the upstream request? What scenarios do you see where this _wouldn't_ work?

**Comment by osoykan** [comment]
Checking the `Content-Length` is a solution in Production but not a solution in testing since `TestServer` `HttpContextBuilder` [doesn't set the ContentLength property,](https://github.com/aspnet/AspNetCore/blob/master/src/Hosting/TestHost/src/ClientHandler.cs#L97-L144) there must be a reason I guess.

So, I think my suggestion would be giving a try to getting into the middle with middleware before the `RunProxy` and set the body to null. @kolomik 

```csharp
_builder
    .Configure(app =>
    {
        app.Use((context, next) =>
        {
            if (context.Request.Method == "GET")
            {
                context.Request.Body = Stream.Null;
            }

            return next();
        });
        app.RunProxy(
            context => context
                .ForwardTo("http://localhost:5000/bar/")
                .Send());
    })
```

After second thought about enhancing `ProxyOptions` also opens a window for the future `Configurable` things which is a bit not good to be flexible on partial parts of the library in my opinion.

**Comment by kolomik** [comment]
Hello @osoykan,

I've updated tests to pass or not to pass original **Content** to Upstream request based on **ContentLength** property. Not sure if this correct approach or not, but here is the PR - https://github.com/damianh/ProxyKit/pull/189

**Comment by damianh** [comment]
I think this is the correct approach. It is basically doing what ProxyKit did previously for _all_ get requests.

@kolomik I've merged the PR and shipped version [2.2.2 to nuget.org](kolomik) thank you!

**Comment by markgould** [comment]
This appears to be an issue again - 2.3.0 works but 2.3.2 is broken.

Looks like this commit broke it: https://github.com/ProxyKit/ProxyKit/commit/09505c720fcae06fb08c4e50ddb23ddcd7fbabc3#diff-7ae9aa92fe8fba004ba2b80d2ee04852
